File: examples/plugins/js/mangadex/chapters.js
// chapters.js — MangaDex get_chapter_list ABI function.

function getChapterList(arg) {
    var mangaID = JSON.parse(arg);
    if (!mangaID) return JSON.stringify([]);

    log.info("mangadex chapters: id=" + mangaID);

    var all = [];
    var offset = 0;
    var total = 999999;

    while (offset < total) {
        var qs = "limit=500&offset=" + offset;
        qs += "&translatedLanguage[]=" + LANG;
        qs += "&order[volume]=asc&order[chapter]=asc";
        qs += "&includes[]=scanlation_group";
        qs += "&includeEmptyPages=0";
        qs += "&" + contentRatingParams();

        var resp = httpGet(API_URL + "/manga/" + mangaID + "/feed?" + qs);
        if (!resp || resp.status < 200 || resp.status >= 300) break;

        var body;
        try {
            body = JSON.parse(resp.body);
        } catch (e) {
            break;
        }

        if (body.total !== undefined) total = body.total;
        if (!body.data || body.data.length === 0) break;

        for (var i = 0; i < body.data.length; i++) {
            var cd = body.data[i];
            var a = cd.attributes;
            if (a.externalURL) continue;

            var chNum = parseFloatSafe(a.chapter);
            var volNum = parseFloatSafe(a.volume);

            var chTitle = "Chapter " + a.chapter;
            if (a.title) chTitle = a.title;
            if (!a.chapter && !a.title) chTitle = "Oneshot";

            all.push({
                id: cd.id,
                manga_id: mangaID,
                title: chTitle,
                chapter_num: chNum,
                volume_num: volNum,
                released_at: parseTime(a.publishAt),
                url: "https://mangadex.org/chapter/" + cd.id,
            });
        }

        offset += body.data.length;
        if (body.data.length < 500) break;
    }

    log.info("mangadex chapters: found " + all.length + " chapters for " + mangaID);
    return JSON.stringify(all);
}

File: examples/plugins/wasm/mangadex/chapters.go
//go:build wasip1

package main

import (
	"encoding/json"
	"goisekai/pkg/types"
	"math"
	"net/url"
	"strconv"

	"github.com/extism/go-pdk"
)

// GetChapterList returns all chapters for a manga (paginated feed).
//
//go:wasmexport GetChapterList
func GetChapterList() int32 {
	var mangaID string
	_ = json.Unmarshal(pdk.Input(), &mangaID)
	if mangaID == "" {
		b, _ := json.Marshal([]types.Chapter{})
		pdk.Output(b)
		return 0
	}

	var all []types.Chapter
	offset := 0
	for {
		q := url.Values{}
		q.Set("limit", "500")
		q.Set("offset", strconv.Itoa(offset))
		q.Add("translatedLanguage[]", lang)
		q.Add("order[volume]", "asc")
		q.Add("order[chapter]", "asc")
		q.Add("includes[]", "scanlation_group")
		q.Set("includeEmptyPages", "0")
		contentRatingQuery(q)

		resp, err := doFetch(apiURL + "/manga/" + mangaID + "/feed?" + q.Encode())
		if err != nil || resp.Status < 200 || resp.Status >= 300 {
			break
		}

		var list chapterListResp
		if err := json.Unmarshal([]byte(resp.Body), &list); err != nil {
			break
		}

		for _, cd := range list.Data {
			if cd.Attributes.ExternalURL != "" {
				continue
			}

			chNum := parseFloat64(cd.Attributes.Chapter)
			if math.IsNaN(chNum) {
				chNum = 0
			}

			volNum := parseFloat64(cd.Attributes.Volume)
			if math.IsNaN(volNum) {
				volNum = 0
			}

			chTitle := "Chapter " + cd.Attributes.Chapter
			if cd.Attributes.Title != "" {
				chTitle = cd.Attributes.Title
			}
			if cd.Attributes.Chapter == "" && cd.Attributes.Title == "" {
				chTitle = "Oneshot"
			}

			all = append(all, types.Chapter{
				ID:         cd.ID,
				MangaID:    mangaID,
				Title:      chTitle,
				ChapterNum: chNum,
				VolumeNum:  volNum,
				ReleasedAt: parseTime(cd.Attributes.PublishAt),
				URL:        "https://mangadex.org/chapter/" + cd.ID,
			})
		}

		offset += list.Limit
		if offset >= list.Total {
			break
		}
	}
	b, _ := json.Marshal(all)
	pdk.Output(b)
	return 0
}
